src/VisualStudio/Core/Def/Implementation/Library/VsNavInfo/NavInfoFactory.cs

```diff
@@ -0,0 +1,265 @@
+            {
+                if (typeSymbol.SpecialType == SpecialType.System_Nullable_T)
+                {
+                    typeSymbol = ((INamedTypeSymbol)typeSymbol).TypeArguments[0];
+                }
+                else if (typeSymbol.TypeKind == TypeKind.Pointer)
+                {
+                    typeSymbol = ((IPointerTypeSymbol)typeSymbol).PointedAtType;
+                }
+                else if (typeSymbol.TypeKind == TypeKind.Array)
+                {
+                    typeSymbol = ((IArrayTypeSymbol)typeSymbol).ElementType;
+                }
+                else
+                {
+                    break;
+                }
+            }
+
+            typeSymbol = typeSymbol.OriginalDefinition;
+
+            if (typeSymbol.TypeKind == TypeKind.Error ||
+                typeSymbol.TypeKind == TypeKind.Unknown ||
+                typeSymbol.TypeKind == TypeKind.Dynamic ||
+                typeSymbol.TypeKind == TypeKind.TypeParameter)
+            {
+                return null;
+            }
+
+            return Create(
+                typeSymbol.ContainingAssembly,
+                project,
+                compilation,
+                useExpandedHierarchy,
+                namespaceName: GetNamespaceName(typeSymbol.ContainingNamespace),
+                className: GetClassName(typeSymbol));
+        }
+
+        public IVsNavInfo CreateForMember(ISymbol memberSymbol, Project project, Compilation compilation, bool useExpandedHierarchy = false)
+        {
+            return Create(
+                memberSymbol.ContainingAssembly,
+                project,
+                compilation,
+                useExpandedHierarchy,
+                namespaceName: GetNamespaceName(memberSymbol.ContainingNamespace),
+                className: GetClassName(memberSymbol.ContainingType),
+                memberName: GetMemberName(memberSymbol));
+        }
+
+        private IVsNavInfo Create(IAssemblySymbol containingAssembly, Project project, Compilation compilation, bool useExpandedHierarchy = false,
+            string namespaceName = null, string className = null, string memberName = null)
+        {
+            // useExpandedHierarchy is true when references are nested inside the project by the
+            // hierarchy. In Class View, they are nested in the Project References node. In Object Browser,
+            // they are not.
+            //
+            // In the case that references are nested inside of the project, we need to create the nav info
+            // differently:
+            //
+            //     project -> containing assembly -> namespace -> type
+            //
+            // Otherwise, we create it like so:
+            //
+            //     containing assembly -> namespace -> type
+
+            string libraryName;
+            string referenceOwnerName = null;
+
+            var isCompilationAssembly = containingAssembly.Identity.Equals(compilation.Assembly.Identity);
```

Comment:
I'm not sure. How would I test that? Note: this is how the code was before. (https://github.com/dotnet/roslyn/pull/6375/files#diff-d907abf2e0e6b19439a2d9ac1dc8565eL179)